Source/SwiftLintCore/Models/Linter.swift

```diff
@@ -145,6 +147,54 @@ private extension Rule {
                           ruleTime: ruleTime,
                           deprecatedToValidIDPairs: deprecatedToValidIDPairs)
     }
+
+    // Produces one region for each disabled rule identifier
```

Comment:
Perhaps you can add some more documentation and examples here, because I'm having a hard time understanding what's happening and why.

Reply:
Let me have think about what to add. The long-winded explanation is this:

Given these rules:

```
"rule1": [
    "regex": "pattern1"
],
"rule2": [
    "regex": "pattern2"
],
```

and this input:

```
// swiftlint:disable rule1
// swiftlint:disable rule2
let pattern2 = ""
// swiftlint:enable rule2
let pattern1 = ""
// swiftlint:enable rule1
```

The regions of the file will be (`index: start - end disabled rules`)

```
0: 1,27 - 2,26 rule1
1: 2,27 - 4,25 rule1, rule2        rule2 is violated in this region
2: 4,26 - 6,25 rule1                  rule1 is violated in this region
3: 6,26 - EOF  no rules disabled
```

These regions are disjoint/non-overlapping. Our algorithm for detecting superfluous disable violations works fine when the disabled areas are non-overlapping, or non-nested, but in this particular nested case, we will get a false superfluous disable command for `rule1` in region `0` and region `1`.

`decompose` is perhaps not quite the right word - what we do here is rewrite the regions, so that there is a region for each rule. These regions can be overlapping. In this above case, we get:

```
1,27 - 6,25 rule1                rule1 is violated in this region
2,27 - 4,25 rule2               rule2 is violated in this region
6,26 - EOF no rules disabled
```

So we correctly get no superfluous disable command violations in this case. I didn't want to change how we represent regions elsewhere, so we just convert them here for this special case.

Reply:
Okay. Got that. I wonder if we should then generally treat regions like that right from the beginning.

> I didn't want to change how we represent regions elsewhere, so we just convert them here for this special case.

Understandable. We should keep that in mind, however. Looks like we can simplify a few things in this area.